--- xr3100.asm	2014-11-17 13:35:02.081676341 +0800
+++ cppm.asm	2014-11-17 13:33:32.784642376 +0800
@@ -1030,13 +1030,14 @@
 	push	dpl
 	push	psw
 	mov	psw,#0
+	setb	PORT_CH4
 	push	rb0r2
 	push	rb0r3
 	push	rb0r4
 	push	rb0r5
+
 	mov	dptr,#servo_output_state
 	movx	a,@dptr
-	mov	r5,a
 	jnz	servo_pulse_ch2
 servo_pulse_ch1:
 	mov	dptr,#ch1_value
@@ -1052,10 +1053,9 @@
 	mov	a,#1
 	movx	@dptr,a
 	setb	PORT_CH1
-	sjmp	X064b
+	sjmp	servo_end
 ;
 servo_pulse_ch2:
-	mov	a,r5
 	cjne	a,#1,servo_pulse_ch3
 	mov	dptr,#ch2_value
 	movx	a,@dptr
@@ -1071,12 +1071,9 @@
 	movx	@dptr,a
 	clr	PORT_CH1
 	setb	PORT_CH2
-	sjmp	X064d
+	sjmp	servo_end
 ;
 servo_pulse_ch3:
-	mov	dptr,#servo_output_state
-	movx	a,@dptr
-	mov	r5,a
 	cjne	a,#2,servo_pulse_ch4
 	mov	dptr,#ch3_value
 	movx	a,@dptr
@@ -1090,14 +1087,11 @@
 	mov	dptr,#servo_output_state
 	mov	a,#3
 	movx	@dptr,a
-	clr	PORT_CH1
 	clr	PORT_CH2
 	setb	PORT_CH3
-	clr	PORT_CH4
-	sjmp	X0651
+	sjmp	servo_end
 ;
 servo_pulse_ch4:
-	mov	a,r5
 	cjne	a,#3,all_servos_done
 	mov	dptr,#ch4_value
 	movx	a,@dptr
@@ -1111,22 +1105,17 @@
 	mov	dptr,#servo_output_state
 	mov	a,#4
 	movx	@dptr,a
-	clr	PORT_CH1
-	clr	PORT_CH2
 	clr	PORT_CH3
-	setb	PORT_CH4
-	sjmp	X0651
 ;
 all_servos_done:
 	clr	TCON_tr1
-	clr	PORT_CH1
-X064b:	clr	PORT_CH2
-X064d:	clr	PORT_CH3
-	clr	PORT_CH4
-X0651:	pop	rb0r5
+
+servo_end:
+	pop	rb0r5
 	pop	rb0r4
 	pop	rb0r3
 	pop	rb0r2
+	clr	PORT_CH4
 	pop	psw
 	pop	dpl
 	pop	dph
